code/game/machinery/computer/launchpad_control.dm

```diff
@@ -38,113 +38,180 @@
-			t += "<BR>"
-			t += "<div class='statusDisplay'>Current offset:</div><BR>"
-			t += "<div class='statusDisplay'>[abs(pad.y_offset)] [pad.y_offset > 0 ? "N":"S"]</div><BR>"
-			t += "<div class='statusDisplay'>[abs(pad.x_offset)] [pad.x_offset > 0 ? "E":"W"]</div><BR>"
-
-			t += "<BR><A href='?src=[REF(src)];launch=1;pad=[current_pad]'>Launch</A>"
-			t += " <A href='?src=[REF(src)];pull=1;pad=[current_pad]'>Pull</A>"
-
-	var/datum/browser/popup = new(user, "launchpad", name, 300, 500)
-	popup.set_content(t.Join())
-	popup.open()
-
-/obj/machinery/computer/launchpad/proc/teleport(mob/user, obj/machinery/launchpad/pad)
-	if(QDELETED(pad))
-		to_chat(user, "<span class='warning'>ERROR: Launchpad not responding. Check launchpad integrity.</span>")
-		return
-	if(!pad.isAvailable())
-		to_chat(user, "<span class='warning'>ERROR: Launchpad not operative. Make sure the launchpad is ready and powered.</span>")
-		return
-	pad.doteleport(user, sending)
-
-/obj/machinery/computer/launchpad/Topic(href, href_list)
-	var/obj/machinery/launchpad/pad
-	if(href_list["pad"])
-		pad = get_pad(text2num(href_list["pad"]))
-
+		data["launchpads"] = pad_list
+	else if(screen == "control")
+		if(QDELETED(current_pad) || (current_pad.stat & NOPOWER))
+			data["pad_active"] = FALSE
+			return data
+		data["pad_active"] = TRUE
+		data["pad_name"] = current_pad.display_name
+		data["abs_x"] = abs(current_pad.x_offset)
+		data["abs_y"] = abs(current_pad.y_offset)
+		data["north_south"] = current_pad.y_offset > 0 ? "N":"S"
+		data["east_west"] = current_pad.x_offset > 0 ? "E":"W"
+
+	return data
+
+/obj/machinery/computer/launchpad/ui_act(action, params)
 	if(..())
 		return
-	if(!LAZYLEN(launchpads))
-		updateDialog()
-		return
-
-	if(href_list["choose_pad"])
-		current_pad = text2num(href_list["pad"])
-
-	if(href_list["raisex"])
-		if(pad.x_offset < pad.range)
-			pad.x_offset++
-
-	if(href_list["lowerx"])
-		if(pad.x_offset > (pad.range * -1))
-			pad.x_offset--
-
-	if(href_list["raisey"])
-		if(pad.y_offset < pad.range)
-			pad.y_offset++
-
-	if(href_list["lowery"])
-		if(pad.y_offset > (pad.range * -1))
-			pad.y_offset--
-
-	if(href_list["reset"])
-		pad.y_offset = 0
-		pad.x_offset = 0
-
-	if(href_list["change_name"])
-		var/new_name = stripped_input(usr, "What do you wish to name the launchpad?", "Launchpad", pad.display_name, 15)
-		if(!new_name)
-			return
-		pad.display_name = new_name
-
-	if(href_list["remove"])
-		if(usr && alert(usr, "Are you sure?", "Remove Launchpad", "I'm Sure", "Abort") != "Abort")
-			launchpads -= pad
-
-	if(href_list["launch"])
-		sending = TRUE
-		teleport(usr, pad)
-
-	if(href_list["pull"])
-		sending = FALSE
-		teleport(usr, pad)
-
-	updateDialog()
\ No newline at end of file
+	switch(action)
+		if("return")
+			screen = "select"
+			. = TRUE
+		if("select_pad")
+			current_pad = get_pad(text2num(params["id"]))
+			screen = "control"
+			. = TRUE
+		if("right")
```

Comment:
The redundancy of these cases is causing me physical pain. Add dx/dy params to the act then you can just do it in one go please.